---
title: "RNAseq splicing outliers FRASER"
author: "Carlos Ruiz"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    number_sections: true
    toc: yes
    fig_caption: yes
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

```{r, echo = FALSE}
library(FRASER)
library(cowplot)
library(ggplot2)
library(dplyr)

splicing_vers <- "v1/"

splicing_folder <- paste0("../results/RNAseq/FRASER_splicing_aberrations/", splicing_vers)

load(paste0(splicing_folder, "FRASER_results.Rdata"))
workingDir(fdsres) <- splicing_folder
```

This report contains an exploration of the outlier results obtained with FRASER.

# General analysis

We removed those aberrations detected in mitochondrial chromosome and collapsed ranges from different measures. 

```{r}
res <- results(fdsres)
res <- res[seqnames(res) != "MT"]
mergeRanges <- function(id){
  range <- reduce(subset(res, sampleID == id)) 
  df <- data.frame(range = as.character(range), 
                   id = id)
  df
}
map <- Reduce(rbind, lapply(unique(res$sampleID), mergeRanges ))
map
table(map$id)
```

22 aberrant splicing events were identified using FRASER. Interestingly, all events but one were detected in patient samples. We will describe these events for each sample. In the following plots, sample with an aberration is plotted separateley and reamining case and control samples are plotted together. 

## 16B12251

Sample 16B12251 presented one aberration, consisting on skipping of exon 7 of MRSP15 gene (chr1, 37Mb). 

![16B12251](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/16B12251.1.png)

## G4CS1CO

Sample G4CS1CO presented one aberration, consisting on retention of intron 1 and use splicing site of another NME2 transcript (chr17, 49Mb). 

![G4CS1CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G4CS1CO.1.png)

## G2CS4CO

Sample G2CS4CO presented two aberrations.

The first aberration was an alternative splicing site in exon 3 of C19orf25 (chr19, 1Mb). 

![G2CS4CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS4CO.1.png)

The second aberration was skipping of exons 5 and 6 of TMLHE (chrX, 154Mb). 

![G2CS4CO.2](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS4CO.2.png)

## G2CS6CO

Sample G2CS6CO presented two aberrations.

The first aberration was skipping of exons 2 to 6 of MNAT1 (chr14, 61Mb). 

![G2CS6CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS6CO.1.png)

The second aberration was skipping of exon 3 of ZFAS1 (chr20, 47Mb).
![G2CS6CO.2](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS6CO.2.png)

## G2CS13CO

Sample G2CS13CO presented three aberrations.

The first aberration was retention of intron 1 of DGCR6 (chr22, 18Mb). 

![G2CS13CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS13CO.1.png)

The second aberration was partial retention of intron 23 of FLNC (chr7, 128Mb). 

![G2CS13CO.2](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS13CO.2.png)


The third aberration was partial retention of intron 4 of AGPAT2 (chr9, 139Mb). 

![G2CS13CO.3](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS13CO.3.png)

## G3CS9CO

Control sample G3CS9CO presented a retention of intron 4 of NME3 (chr16, 1Mb). 

![G3CS9CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G3CS9CO.1.png)

![G3CS9CO.2](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G3CS9CO.2.png)

## G4CS13CO

Sample G4CS13CO presented an alternative splicing stie for exon 14 (or 6 depending on transcript) of RPS6KB2 (chr11, 67Mb). 

![G4CS13CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G4CS13CO.1.png)


## G2CS9CO

Sample G2CS9CO presented higher coverage in a region were no genes are annotated in genome browser. This might indicate the presence of a different non-coding RNA.  

![G2CS9CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS9CO.1.png)


## G5CS6CO

Sample G5CS6CO presented the skipping of exon 2 of SNHG32 (chr6, 31Mb)

![G5CS6CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G5CS6CO.1.png)

## G2CS8CO

Sample G2CS8CO presented six aberrations.

The first aberration was retention of intron 7 (or 99, depending on the transcript) of RYR2 (chr1, 237Mb). 

![G2CS8CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS8CO.1.png)

The second aberration was retention of intron 4 (or 11 or 12, depending on the transcript) of DENND5A (chr11, 9Mb). 

![G2CS8CO.2](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS8CO.2.png)

The third aberration was an alternative splicing site of exon 11 (or 18, 19, 21, depending on the transcript) of USP28 (chr11, 113Mb). 

![G2CS8CO.3](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS8CO.3.png)


The fourth aberration was an alternative splicing site of exon 3 (or 2, depending on the transcript) of RNF10 (chr12, 120Mb). 

![G2CS8CO.4](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS8CO.4.png)

The fifth aberration was retention of intron 21 (or 18, 20 , depending on the transcript) of MARCH6 (chr5, 10Mb). 

![G2CS8CO.5](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS8CO.5.png)

The sixth aberration was alternative splicing site of exon 1 of DNAJA1 (chr9, 33Mb). 

![G2CS8CO.6](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G2CS8CO.6.png)


## G4CS3CO

Sample G4CS3CO presented two aberrations.

The first aberration was increased covarege in last exon (exon 10) of MIIP (chr1, 12Mb).

![G4CS3CO](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G4CS3CO.1.png)

The second aberration was alternative splicing site in exon 4 of REXO2 (chr11, 114Mb).

![G4CS3CO.2](../results/RNAseq/FRASER_splicing_aberrations_plots/v1/G4CS3CO.2.png)